dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/registry/ServerNodeManager.java

```diff
@@ -210,40 +213,59 @@ class MasterDataListener implements SubscribeListener {
         public void notify(Event event) {
             final String path = event.path();
             final Type type = event.type();
+            final String data = event.data();
             if (registryClient.isMasterPath(path)) {
                 try {
+                    String[] parts = path.split("/");
+                    final String masterAddress = parts[parts.length - 1];
                     if (type.equals(Type.ADD)) {
                         log.info("master node : {} added.", path);
                         updateMasterNodes();
-                    }
-                    if (type.equals(Type.REMOVE)) {
+                    } else if (type.equals(Type.REMOVE)) {
                         log.info("master node : {} down.", path);
                         updateMasterNodes();
                         alertDao.sendServerStoppedAlert(1, path, "MASTER");
+                    } else if (type == Type.UPDATE) {
```

Comment:
It's better don't change this listener, we need to cache all master heartbeat info, but only need to filter the overload master when calculate slot, so it's better to split this class to ServerNodeManager and MasterSlotManager. ServerNodeManager only need to synchronize the data from the registry.

Reply:
Okay, I will move all slot-related info to `MasterSlotManager`, but it seems that whenever we receive `ADD/UPDATE/DELETE` event, we all have to update update master heartbeat info and call `MasterSlotManager` to update slot.